tools/ports/__init__.py

```diff
@@ -31,39 +31,49 @@
 logger = logging.getLogger('ports')
 
 
+def load_port(name):
+  expected_attrs = ['get', 'clear', 'show']
+  port = __import__(name, globals(), level=1)
+  ports.append(port)
+  port.name = name
+  ports_by_name[port.name] = port
+  for a in expected_attrs:
+    assert hasattr(port, a), 'port %s is missing %s' % (port, a)
+  if not hasattr(port, 'needed'):
+    port.needed = lambda s: name in s.PORTS
+  else:
+    needed = port.needed
+    port.needed = lambda s: needed(s) or name in s.PORTS
+  if not hasattr(port, 'process_dependencies'):
+    port.process_dependencies = lambda x: 0
+  if not hasattr(port, 'linker_setup'):
+    port.linker_setup = lambda x, y: 0
+  if not hasattr(port, 'deps'):
+    port.deps = []
+  if not hasattr(port, 'process_args'):
+    port.process_args = lambda x: []
+  if not hasattr(port, 'variants'):
+    # port variants (default: no variants)
+    port.variants = {}
+
+  for variant, extra_settings in port.variants.items():
+    if variant in port_variants:
+      utils.exit_with_error('duplicate port variant: %s' % variant)
+    port_variants[variant] = (port.name, extra_settings)
+
+
 @ToolchainProfiler.profile()
 def read_ports():
-  expected_attrs = ['get', 'clear', 'show', 'needed']
   for filename in os.listdir(ports_dir):
     if not filename.endswith('.py') or filename == '__init__.py':
       continue
     filename = os.path.splitext(filename)[0]
-    port = __import__(filename, globals(), level=1)
-    ports.append(port)
-    port.name = filename
-    ports_by_name[port.name] = port
-    for a in expected_attrs:
-      assert hasattr(port, a), 'port %s is missing %s' % (port, a)
-    if not hasattr(port, 'process_dependencies'):
-      port.process_dependencies = lambda x: 0
-    if not hasattr(port, 'linker_setup'):
-      port.linker_setup = lambda x, y: 0
-    if not hasattr(port, 'deps'):
-      port.deps = []
-    if not hasattr(port, 'process_args'):
-      port.process_args = lambda x: []
-    if not hasattr(port, 'variants'):
-      # port variants (default: no variants)
-      port.variants = {}
-
-    for variant, extra_settings in port.variants.items():
-      if variant in port_variants:
-        utils.exit_with_error('duplicate port variant: %s' % variant)
-      port_variants[variant] = (port.name, extra_settings)
-
-  for dep in port.deps:
-    if dep not in ports_by_name:
-      utils.exit_with_error('unknown dependency in port: %s' % dep)
+    load_port(filename)
+
+  for port in ports:
+    for dep in port.deps:
+      if dep not in ports_by_name:
+        utils.exit_with_error('unknown dependency in port: %s' % dep)
```

Comment:
Interesting.. this looks like a bug fix, is that right?  That old code was only checking the deps of the final port that it read?

Feel free to split this out and if you like (i.e. no need to block this bugfix on the rest of the discussion in this PR).

Reply:
Yes it is a bug fix and I mentioned that I fixed it in my description of the PR...

Reply:
Since this is an internal check (which IMO should not be done everytime ports are read but in a test), there is no chance that something will break unless another port were to be added with invalid dependencies which seems unlikely. As a result I don't see the need to extract this into its own PR...